internal-fs/opt/kexec/check_kexec

```diff
@@ -26,6 +26,24 @@ elif [ -f /mnt/sdcard/x1plus/printers/${BBL_SN}/quick-boot ]; then
     QUICKBOOT_FLAG="true"
 fi
 
+# Also check, and get/clear our OTA filename flag for auto install
+if [ -f /mnt/sdcard/x1plus/printers/${BBL_SN}/settings.json ]; then
+    OTA_FILENAME=$(/opt/kexec/jq -r ".\"ota.filename\" // null" /mnt/sdcard/x1plus/printers/${BBL_SN}/settings.json)
+    # Did we have something set?
+    if [ "${OTA_FILENAME}" != "null" ]; then
+        if [ -f /mnt/sdcard/${OTA_FILENAME} ]; then
+            export X1P_OTA=${OTA_FILENAME}
+            QUICKBOOT_FLAG="false" # Back to false, so we make it to installer if we are re-installing a matching named build
+        fi
+        # Always clear the flag no matter what if it wasn't null, we have to do this ourselves as we don't have dbus
+        # for x1plusd as we are NOT in x1plus yet!
+        TMPSETTINGS=$(mktemp)
+        /opt/kexec/jq ".\"ota.filename\" = null" /mnt/sdcard/x1plus/printers/${BBL_SN}/settings.json > "${TMPSETTINGS}"
+        mv "${TMPSETTINGS}" /mnt/sdcard/x1plus/printers/${BBL_SN}/settings.json 2> /dev/null # suppress permissions warning
```

Comment:
this cross filesystem move does not update things atomically the way you want it to: it unlinks the settings.json before moving the new one in place.  write to a temporary file in the SD card, *then* sync, *then* mv it in place over the top of the old one, *then* sync again

Reply:
I think there's a misunderstanding; jq isn't removing the file, it's modifying the output of it and saving it to tmp. We then move it ontop of the existing settings.json and syncing. Because of this there's never a moment where the file doesn't exist; this is the same logic as emmc_dump uses to clear it's flag.

Reply:
Arguably, emmc_dump is wrong then, too :-)

The problem is not in the `jq` but in the `mv`.  In this case, the syscalls that `mv` executes, morally, are as follows:

```
open(TMPSETTINGS, O_RDONLY) -> fd_in
open("/mnt/sdcard/.../settings.json", O_WRONLY | O_CREAT | O_TRUNC) -> fd_out
read(fd_in, ...)
write(fd_out, ...)
close(fd_in)
close(fd_out)
```

This is not atomic: after the `O_TRUNC` open, if power is lost or the system crashes, settings.json ends up zero bytes.  What you want is something that uses the `rename` syscall, which, unfortunately, does not work cross-filesystem; from `rename`'s man page: "If newpath already exists, it will be atomically replaced, so that there is no point at which another process attempting to access newpath will find it missing."  (We do this already in [interpose.cpp](https://github.com/X1Plus/X1Plus/blob/main/bbl_screen-patch/interpose.cpp#L256-L259).)

The good news is that there is a tool that uses the `rename` syscall!  Indeed, that tool is ... `mv`.  The catch is that the source and destination need to be on the same filesystem.

So what you probably want to do is something like:

```
jq ... /mnt/sdcard/.../settings.json > /mnt/sdcard/.../settings.json.tmp
sync # make sure settings.json.tmp is fully committed to disk before we start writing the directory entry
mv /mnt/sdcard/.../settings.json.tmp /mnt/sdcard/.../settings.json
sync # make sure the directory entry is committed to disk
```